src/ProjectFileTools.MSBuild/Worksapce.cs

```diff
@@ -0,0 +1,149 @@
+        public string ResolveDefinition(string filePath, string sourceText, int position)
+        {
+            String file = null;
+
+            if (_project != null)
+            {
+                XmlDocumentSyntax root = Parser.ParseText(sourceText);
+                SyntaxNode syntaxNode = SyntaxLocator.FindNode(root, position);
+
+                if (syntaxNode.ParentElement.Name.Equals("Import"))
+                {
+                    while (syntaxNode.Parent.ParentElement == syntaxNode.ParentElement)
+                    {
+                        syntaxNode = syntaxNode.Parent;
+                    }
+
+                    int nodeStart = syntaxNode.Parent.Span.Start;
+                    int col = nodeStart - Utilities.GetStartOfLine(sourceText, nodeStart) + 1;
+                    int line = Utilities.GetLine(sourceText, nodeStart) + 1;
+
+                    foreach (ResolvedImport import in _project.Imports)
+                    {
+                        ElementLocation location = import.ImportingElement.Location;
+                        if (location.File == filePath && col == location.Column && line == location.Line)
+                        {
+                            file = import.ImportedProject.FullPath;
+                            break;
+                        }
+                    }
+                }
+            }
+            return file;
+        }
+
+        internal bool ContainsProject(string filePath)
+        {
+            ReloadIfNecessary();
+            return _containedFiles.Contains(filePath);
+        }
+
+        private void ReloadIfNecessary()
+        {
+            if (_needsReload)
+            {
+                string tempPath = _project.FullPath;
+                try
+                {
+                    ProjectCollection tempCollection = new ProjectCollection();
+                    Project tempProject = tempCollection.LoadProject(tempPath);
+                    _collection.UnloadAllProjects();
+                    _collection = tempCollection;
+                    _project = tempProject;
+                    UpdateContainedFiles();
+                }
+                // TODO: Propagate error to the errors list
+                catch
+                {
+
+                }
+
+                _needsReload = false;
+            }
+        }
+
+        private void UpdateContainedFiles()
+        {
+            _containedFiles.Clear();
+
+            foreach(FileSystemWatcher watcher in _watchers)
+            {
+                watcher.Changed -= MarkReload;
+                watcher.Deleted -= MarkReload;
+                watcher.Renamed -= MarkReload;
+                watcher.Dispose();
+            }
+            _watchers.Clear();
+
+            foreach (ResolvedImport import in _project.Imports)
+            {
+                _containedFiles.Add(import.ImportingElement.Location.File);
+                _containedFiles.Add(import.ImportedProject.FullPath);
```

Comment:
I think `_containedFiles.Add(import.ImportingElement.Location.File);` is unnecessary. You do however need to add `_project.FullPath` separately.